From 278faf34a9543140a8511f48bdf9a5c5df60ffd3 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:01:27 -0400
Subject: [usb] serial: call handle_dcd_change in ftdi driver

Message-id: <1410887463-20674-16-git-send-email-dzickus@redhat.com>
Patchwork-id: 95371
O-Subject: [RHEL7 PATCH 015/591] USB: serial: call handle_dcd_change in ftdi driver.
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit d14654dff7a3520b5220367b848732a0a8ccdabe
Author: Paul Chavent <Paul.Chavent@onera.fr>
Date:   Mon Sep 16 08:40:59 2013 +0200

    USB: serial: call handle_dcd_change in ftdi driver.

    When the device receive a DCD status change, forward the signal to the
    USB serial system. This way, we can detect, for instance, PPS pulses.

    Signed-off-by: Paul Chavent <paul.chavent@onera.fr>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/serial/ftdi_sio.c b/drivers/usb/serial/ftdi_sio.c
index dccb4db..fb0d537 100644
--- a/drivers/usb/serial/ftdi_sio.c
+++ b/drivers/usb/serial/ftdi_sio.c
@@ -1967,8 +1967,16 @@ static int ftdi_process_packet(struct usb_serial_port *port,
    port->icount.dsr++;
   if (diff_status & FTDI_RS0_RI)
    port->icount.rng++;
-  if (diff_status & FTDI_RS0_RLSD)
+  if (diff_status & FTDI_RS0_RLSD) {
+   struct tty_struct *tty;
+
    port->icount.dcd++;
+   tty = tty_port_tty_get(&port->port);
+   if (tty)
+    usb_serial_handle_dcd_change(port, tty,
+      status & FTDI_RS0_RLSD);
+   tty_kref_put(tty);
+  }
 
   wake_up_interruptible(&port->port.delta_msr_wait);
   priv->prev_status = status;
-- 
1.7.1